larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h

```diff
@@ -0,0 +1,393 @@
+     *
+     *  @param  index the index in the probability vector
+     */
+    void VerifyElementRequest(const size_t index) const;
+
+    float m_xUpperBound;                                              ///< the upper bound of the probability vector
+    bool m_useWidths;                                                 ///< controls whether bin widths are used in calculations
+    DiscreteProbabilityData m_discreteProbabilityData;                ///< the probability data
+};
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline size_t DiscreteProbabilityVector::GetSize() const
+{
+    return m_discreteProbabilityData.size();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline float DiscreteProbabilityVector::GetX(const size_t index) const
+{
+    this->VerifyElementRequest(index);
+
+    return m_discreteProbabilityData.at(index).GetX();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline float DiscreteProbabilityVector::GetProbability(const size_t index) const
+{
+    this->VerifyElementRequest(index);
+
+    return m_discreteProbabilityData.at(index).GetDensityDatum()*(m_useWidths ? m_discreteProbabilityData.at(index).GetWidth() : 1.f);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline float DiscreteProbabilityVector::GetProbabilityDensity(const size_t index) const
+{
+    this->VerifyElementRequest(index);
+
+    return m_discreteProbabilityData.at(index).GetDensityDatum();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline float DiscreteProbabilityVector::GetCumulativeProbability(const size_t index) const
+{
+    this->VerifyElementRequest(index);
+
+    return m_discreteProbabilityData.at(index).GetCumulativeDatum();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline float DiscreteProbabilityVector::GetWidth(const size_t index) const
+{
+    this->VerifyElementRequest(index);
+
+    return m_discreteProbabilityData.at(index).GetWidth();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline void DiscreteProbabilityVector::GetAllAtIndex(const size_t index, float &x, float &probabilityDensity,
+    float &cumulativeProbability, float &width) const
+{
+    this->VerifyElementRequest(index);
+
+    const DiscreteProbabilityDatum &theDatum(m_discreteProbabilityData.at(index));
```

Comment:
Possible to just return theDatum, or want to decouple by using floats. This is fine, but won't scale well with addition of any further info.

Reply:
I mentioned the reasoning in another PR comment.  The function was made for convenience and returning the floats individually, without having to unpack a struct is maximally convenient.
I do 100% agree that if the number of members in the datum grows then this becomes a problem.